### DISCUSSION TITLE
Diverging code path for 32 bit & 64 bit

### DISCUSSION THREAD

**Discussion by x86fr** [memtest86plus/memtest86plus#305] [General]
We currently support computers over a period of 30 years, with barely the same code compiled on 32 and 64 bit

This will become more and more difficult to handle, especially if we want to add more "modern" features like SIMD tests.

At some point, we will have to:

- Keep 32 bit binaries optimized for 32-bit CPU (486/P5 to Pentium 3/Athlon XP)
- Upgrade the 64 bit binary to more modern instructions set (march=p4 / -msse2)

An objection to do this (there is no x86-64 CPU without SSE2)? 
Should we try to keep both code in the same files (with big ifdef/ifndef) or split 32 and 64bit code in different files when needed (like for startup32 and startup64)?

**Comment by frispete** [comment]
I would suggest splitting it up. While physical 32-bit CPUs are mostly retired, the 64-bit world gets increasingly complicated.

We're facing this in our distributions. https://en.wikipedia.org/wiki/X86-64 just describes the top of the iceberg.

Of course, this project will always stick with the most conservative approach in selecting the target architecture, but
maybe some kind soul will implement additional checks based on the AVX-256 SIMD instruction set one day. As far as I know,
this is implemented in separate areas on-die, and therefore exercising these areas would result in extended coverage of 
memory <-> cpu transfers, making memtest86+ even more valuable. 

For basic x86-64 targets, it is safe to assume SSE2 support nowadays (but I'm in no way a µarch specialist). According to 
https://en.wikipedia.org/wiki/SSE2, models that do not support SSE2 are more than 20 years old (specifically: AMD +20y, 
Intel +23y).

**Reply by martinwhitaker** [reply]
I don't have any objection to changing the build flags, but will we actually gain anything? The only place we care about performance is in the inner loops of the memory tests themselves. And I would have expected that with SMP enabled, most CPUs are capable of max'ing out the memory bandwidth even with the current code.

We'd need to check that enabling SIMD instructions doesn't require any changes in the interrupt handling code.

**Reply by debrouxl** [reply]
We know that in a general-purpose kernel with task switching, SIMD registers require handling in the task switching code. Thankfully, in memtest86+, we don't *currently* have to care about task switching.
startup64.S contains a disabled section of code for enabling SSE.

**Reply by martinwhitaker** [reply]
> startup64.S contains a disabled section of code for enabling SSE.

Yes, and I can't now remember whether I disabled because it wasn't needed, or because it caused problems...

**Reply by x86fr** [reply]
128-bit stores from SIMD registers may trigger some new memory faults (or show them faster). 
It's highly speculative and real-world tests must be performed of course.

**Comment by x86fr** [comment]
With the introduction of X86S, we will have no choice but making a different 64-bit binary. :-/
More info: https://cdrdv2.intel.com/v1/dl/getContent/776648

**Reply by martinwhitaker** [reply]
> With the introduction of X86S, we will have no choice but making a different 64-bit binary. :-/ 

From a very brief read, I don't see why this should cause us any problems. Such systems will surely be UEFI-only, which already bypasses all the legacy boot stuff in memtest86+ when booting the BSP. AP boot will presumably just need to vector to the right place in startup64. We intend to add x2APIC support anyway.

**Reply by x86fr** [reply]
Yes indeed. I was thinking about legacy 64-bit (without UEFI) but for sure, they will be UEFI only.

Beside from this, what do you think about compiling the 64-bit for SSE2 target?

**Reply by debrouxl** [reply]
Yeah, X86-S will require adding a different binary anyway, so we'll end up having at least 5 binaries:
* a 32-bit BIOS binary, down compatible to 586-class or presumably eventually 486-class ( #120 ) processors, which does not preclude conditional usage of code compatible with newer processors at select places;
* a 32-bit UEFI binary, since per e.g. #243 , there are 32-bit-only UEFI systems significantly newer than the introduction of UEFI circa 2005-2006. This one could have a higher instruction set baseline, since UEFI is quite a bit newer than Pentium;
* a 64-bit BIOS binary for x86-64, since there are non-UEFI 64-bit systems. This one can use SSE2 since this is architectural in AMD64 / EM64T, and could conditionally use newer SIMD instructions at places;
* a 64-bit UEFI binary for x86-64, which can use SSE2 and conditionally newer SIMD instructions, for the same reason;
* a 64-bit UEFI binary for future X86-S, which can use whichever instruction baseline Intel will end up mandating as part of the final version of the X86-S spec, though there's nothing in that direction in the X86-S EAS v1.0.

Beyond compiler flags, there's room for further divergence of the code paths depending on the platform, even without littering the code with too many `#ifdef`. For instance:
* no need for the x86_64 binaries to contain support for temperature, SMBus, IMC, EDAC, etc. support for stuff which exists only on 32-bit platforms. Say, is there a single x86-64 processor which supports SDRAM ?
* no need for the X86-S binary to contain hardware detection support for earliers chipsets, processors and associated RAM generations;
* 5-level paging support could be specific to x86-64 and X86-S. However, it's clear that 4-level paging already gives access to amounts of memory much larger than current memtest86+ can test in an acceptable amount of time, even with the speedup brought by the initial NUMA awareness code which reduces cross-NUMA node access bottlenecks, so it's not like we're going to add 5-level paging support to memtest86+ any time soon :)

Large amounts of memory can be tested at higher bandwidth by using one set of page directories and page tables per NUMA node, and having independent, parallel testing across all NUMA nodes (as opposed to the current dependent, parallel testing across all NUMA nodes, or the WIP parallel testing inside a single NUMA node), but that requires a sizable rework of the way memtest86+ works, both low-level layers and UI layers, and is off-topic for this discussion anyway...

**Reply by x86fr** [reply]
Maybe in some times, the 32-bit UEFI binary does not need to be maintained. UEFI-only 32-bit system are quite rare and they're all EOL since 10+ years. BIOS64 is more common, but 32-bit binary w/PAE is also fine on these system to test > 4GB. Finally, as Martin said, UEFI64 X86S could probably be included in the same UEFI64 binary. 

So I think only 2 binaries (UEFI64/SSE2 and BIOS32/i586) will cover 99+% of users in the future. 

Agree with 5-level paging.

PS: I don't understand why you said
> no need for the x86_64 binaries to contain support for temperature, SMBus, IMC, EDAC, etc. support for stuff which exists only on 32-bit platforms.

All of them are supported on both 32 & 64 bit as they're unrelated to CPU ISA.

**Reply by debrouxl** [reply]
Well, a BIOS 32-bit PAE binary isn't enough for years of 64-bit-capable but definitely / sometimes BIOS-only Opteron or Xeon systems. Even some late K8 systems support >> 64 GB of DDR2 (amusingly, I found about the likes of Sun Fire X4440 and X4640 earlier today, unrelated to this discussion), let alone K10 or Bulldozer systems: my RS904A servers support 32 x 32 GB = 1 TB of DDR3 and are now equipped with 128 GB, down from 256 GB, to speed up memtest86+ tests.
At least, the 32-bit UEFI binary should go away before the 64-bit BIOS binary, and there's indeed no rush deleting things which should be zero maintenance now that they're done and remain unchanged.

For now, I wouldn't yet count on Intel making it possible to support x86-64 and X86-S in the same binary, though it may be interesting for some use cases if they do :)
All x86-64 bare metal software attempting to support X86-S will require an update anyway, no matter whether it is made possible to support both ISA variants in the same binary.

I'll give several examples to try and show what I meant about the lack of a need for x86_64 binaries to contain support for 32-bit-only stuff:
* since even the oldest K8 use DDR, and (AFAICT) all of the oldest 64-bit P4 use socket LGA775 which, unlike Socket 478, only supports DDR (2 & 3, right ?), I don't expect that any AMD or Intel x86-64 processor uses SDRAM, and unless one of their competitors did that major blunder, the SDRAM SPD decoder ought to be effectively dead code in a 64-bit binary. Possibly likewise for the RDRAM SPD decoder. Should these bits be confirmed dead code in a 64-bit binary, a couple `#ifdef` + `#endif` in `system/smbus.c` would get rid of them, at zero maintenance cost;
* if someone eventually contributes IMC support for Athlon XP, or 32-bit-only P4, the corresponding bits, possibly a strict subset thereof if the implementation shares some code with the implementation of IMC support for newer platforms, would be dead code in a 64-bit binary;
* some of the PCI IDs from VIA, ALi, Nvidia or Serverworks have certainly never been put on a MB with an x86_64 processor, so support code for them is technically dead code in a 64-bit binary. However, don't worry, for such fine-grained pieces, I don't plan on spending time trying to build the exact list and littering `system/smbus.c::find_smb_controller()` with `#ifndef __x86_64__` / `#endif` - you rightfully wouldn't merge such a patch even if I foolishly published it, anyway ^^

However, PIIX4 and ICH5 being used by both 32-bit platforms and 64-bit platforms, their supporting code belongs in both binaries, no question about that, even though the wisdom of running 32-bit memtest86+ binaries on RPL / SPR / Zen 4 platforms is, um, *questionable*.
